File: src/memory_vector.py
"""
memory_vector.py

ChromaDB-backed vector store for memory entries.
Shares the EmbeddingClient with RAG to save memory.
Stores pre-computed embeddings (ChromaDB does not manage embedding).
"""

import logging
from pathlib import Path
from typing import List, Dict, Optional

from src.embedding_lanes import (
    LANE_CUSTOM,
    LANE_FASTEMBED,
    build_embedding_lanes,
    collection_name,
    dedupe_results,
    lane_count,
    migrate_legacy_collection,
)

logger = logging.getLogger(__name__)


def _is_missing_collection_error(error: Exception) -> bool:
    """Return whether Chroma reported that the named collection is absent."""
    error_type = type(error)
    return (
        error_type.__module__ == "chromadb.errors"
        and error_type.__name__ in {"NotFoundError", "InvalidCollectionException"}
    )


class MemoryVectorStore:
    """Vector index over memory entries for semantic retrieval."""

    COLLECTION_NAME = "odysseus_memories"
    REBUILD_MARKER = ".memory-vector-rebuild-required"

    def __init__(self, data_dir: str, embedding_model=None):
        self._model = embedding_model
        self._rebuild_marker = Path(data_dir) / self.REBUILD_MARKER
        self._collection = None
        self._lanes = []
        self._healthy = False

        self._initialize()

    def _initialize(self):
        try:
            if self._rebuild_marker.exists():
                # A previous rebuild did not finish durably. Never adopt its
                # non-empty collections as authoritative merely because they
                # survived a process restart. Reset them so startup sees an
                # empty index and rebuilds from the JSON source of truth.
                logger.warning("Discarding incomplete memory-vector rebuild")
                self._replace_collections()
                self._clear_rebuild_marker()
                return

            self._lanes = build_embedding_lanes(self.COLLECTION_NAME)
            if not self._lanes:
                raise RuntimeError("No embedding lanes available")

            self._healthy = True
            self._collection = next(
                (lane.collection for lane in self._lanes if lane.name == LANE_FASTEMBED),
                self._lanes[0].collection,
            )
            migrate_legacy_collection(self.COLLECTION_NAME, self._lanes)
            logger.info(
                "MemoryVectorStore ready (lanes=%s entries=%s)",
                [lane.name for lane in self._lanes],
                self.count(),
            )

        except Exception as e:
            self._healthy = False
            self._lanes = []
            self._collection = None
            logger.error(f"MemoryVectorStore init failed: {e}")

    def _mark_rebuild_incomplete(self) -> None:
        """Durably poison collections until a complete rebuild is committed."""
        self._rebuild_marker.write_text("rebuild required\n", encoding="utf-8")

    def _clear_rebuild_marker(self) -> None:
        self._rebuild_marker.unlink(missing_ok=True)

    def _delete_collection_names(self, names) -> None:
        """Delete named collections, ignoring only explicit not-found errors."""
        from src.chroma_client import get_chroma_client

        client = get_chroma_client()
        for name in names:
            try:
                client.delete_collection(name)
            except Exception as e:
                if _is_missing_collection_error(e):
                    continue
                raise

    @property
    def healthy(self) -> bool:
        return self._healthy

    def _embed(self, texts: List[str]) -> List[List[float]]:
        if not self._lanes:
            return []
        return self._lanes[0].encode(texts)

    def count(self) -> int:
        """Return the number of stored vectors."""
        if not self._healthy:
            return 0
        return lane_count(self._lanes)

    def _collections_for_delete(self):
        collections = []
        seen = set()

        def add(collection) -> None:
            if collection is None:
                return
            key = getattr(collection, "name", None) or id(collection)
            if key in seen:
                return
            seen.add(key)
            collections.append(collection)

        for lane in self._lanes:
            add(lane.collection)

        try:
            from src.chroma_client import get_chroma_client

            client = get_chroma_client()
            for lane_name in (LANE_CUSTOM, LANE_FASTEMBED):
                try:
                    add(client.get_collection(collection_name(self.COLLECTION_NAME, lane_name)))
                except Exception:
                    pass
        except Exception:
            pass

        return collections

    def add(self, memory_id: str, text: str):
        """Add a single memory entry to the vector index."""
        if not self._healthy:
            return
        for lane in self._lanes:
            try:
                existing = lane.collection.get(ids=[memory_id])
                if existing["ids"]:
                    continue
                lane.collection.add(
                    ids=[memory_id],
                    embeddings=lane.encode([text]),
                    documents=[text],
                    metadatas=[{"source": "memory"}],
                )
            except Exception as e:
                logger.warning("memory add failed in %s lane for %s: %s", lane.name, memory_id, e)

    def remove(self, memory_id: str):
        """Remove a memory entry. O(1) — no rebuild needed."""
        if not self._healthy:
            return
        for collection in self._collections_for_delete():
            try:
                collection.delete(ids=[memory_id])
            except Exception as e:
                logger.warning(f"memory remove {memory_id}: {e}")

    def search_with_status(self, query: str, k: int = 8) -> tuple[List[Dict], bool]:
        """Search for the most relevant memory IDs by semantic similarity.
        Return ``(results, query_healthy)`` so callers can distinguish a real
        empty result from a runtime vector outage.

        ChromaDB cosine distance = 1 - cosine_similarity.
        We convert back: similarity = 1.0 - distance.
        """
        if not self._healthy:
            return [], False

        out = []
        successful_lane = False
        lane_priority = {LANE_CUSTOM: 0, LANE_FASTEMBED: 1}
        for lane in self._lanes:
            try:
                if lane.count() == 0:
                    successful_lane = True
                    continue
                results = lane.collection.query(
                    query_embeddings=lane.encode([query]),
                    n_results=min(k, lane.count()),
                    include=["distances"],
                )
                successful_lane = True
                for idx, mid in enumerate(results["ids"][0]):
                    distance = results["distances"][0][idx]
                    out.append({
                        "memory_id": mid,
                        "score": round(1.0 - distance, 4),
                        "embedding_lane": lane.name,
                    })
            except Exception as e:
                logger.warning("memory search failed in %s lane: %s", lane.name, e)
        out.sort(key=lambda row: (-row["score"], lane_priority.get(row["embedding_lane"], 99)))
        return dedupe_results(out, id_key="memory_id", limit=k), successful_lane

    def search(self, query: str, k: int = 8) -> List[Dict]:
        """Compatibility search API returning only result rows."""
        results, _query_healthy = self.search_with_status(query, k=k)
        return results

    def find_similar(self, text: str, threshold: float = 0.92) -> Optional[str]:
        """Check if a near-duplicate exists. Returns memory_id if found, else None."""
        if not self._healthy or self.count() == 0:
            return None

        for lane in self._lanes:
            try:
                if lane.count() == 0:
                    continue
                results = lane.collection.query(
                    query_embeddings=lane.encode([text]),
                    n_results=1,
                    include=["distances"],
                )
                if results["ids"][0]:
                    distance = results["distances"][0][0]
                    similarity = 1.0 - distance
                    if similarity >= threshold:
                        return results["ids"][0][0]
            except Exception as e:
                logger.warning("memory similarity search failed in %s lane: %s", lane.name, e)
        return None

    def _replace_collections(self) -> None:
        """Delete every memory-vector collection and create fresh lanes."""
        names = {
            self.COLLECTION_NAME,
            collection_name(self.COLLECTION_NAME, LANE_CUSTOM),
            collection_name(self.COLLECTION_NAME, LANE_FASTEMBED),
        }
        # Older Chroma clients and the repository's supported lightweight
        # client contract do not expose ``list_collections``.  The collection
        # names are deterministic, so deleting each known name is both more
        # compatible and sufficient to prevent a stale inactive lane (or the
        # legacy collection) from being migrated back into a rebuild.
        self._delete_collection_names(names)

        self._lanes = build_embedding_lanes(self.COLLECTION_NAME)
        if not self._lanes:
            self._healthy = False
            self._collection = None
            raise RuntimeError("No embedding lanes available after memory-vector reset")
        self._collection = next(
            (lane.collection for lane in self._lanes if lane.name == LANE_FASTEMBED),
            self._lanes[0].collection,
        )
        self._healthy = True

    def clear(self, *, strict: bool = False) -> bool:
        """Clear all memory vectors.

        ``strict=True`` is used by persistence transactions: failures propagate
        so SQL/JSON state is not committed while stale vectors remain.
        """
        try:
            self._replace_collections()
            return True
        except Exception as e:
            self._healthy = False
            logger.error("memory vector clear failed: %s", e)
            if strict:
                raise
            return False

    def rebuild(self, memories: List[Dict], *, strict: bool = False) -> bool:
        """Rebuild the entire index from a list of memory entries.
        Each entry must have 'id' and 'text' keys."""
        try:
            # Set this before mutating Chroma. A crash, strict exception, or
            # failed cleanup must remain visible to the next process.
            self._mark_rebuild_incomplete()
            self._replace_collections()
        except Exception as e:
            self._healthy = False
            logger.error("memory rebuild reset failed: %s", e)
            if strict:
                raise
            return False

        texts = []
        ids = []
        for mem in memories:
            text = mem.get("text", "").strip()
            mid = mem.get("id", "")
            if text and mid:
                texts.append(text)
                ids.append(mid)

        if texts:
            # Batch in chunks of 100 to avoid oversized requests
            failed_lanes = set()
            primary_error = None
            for i in range(0, len(texts), 100):
                batch_texts = texts[i:i + 100]
                batch_ids = ids[i:i + 100]
                for lane in self._lanes:
                    if lane.name in failed_lanes:
                        continue
                    try:
                        lane.collection.add(
                            ids=batch_ids,
                            embeddings=lane.encode(batch_texts),
                            documents=batch_texts,
                            metadatas=[{"source": "memory"}] * len(batch_ids),
                        )
                    except Exception as e:
                        failed_lanes.add(lane.name)
                        logger.warning("memory rebuild failed in %s lane: %s", lane.name, e)
                        if strict:
                            primary_error = e
                            break
                if primary_error is not None:
                    break

            if primary_error is not None:
                # Strict mode remains fail-fast, but first make every lane
                # non-authoritative: each may contain earlier successful
                # batches. Preserve the embedding/write exception even if
                # best-effort cleanup also fails; the marker then protects the
                # next process from adopting the residue.
                all_names = {
                    collection_name(self.COLLECTION_NAME, lane.name)
                    for lane in self._lanes
                }
                self._healthy = False
                self._lanes = []
                self._collection = None
                try:
                    self._delete_collection_names(all_names)
                except Exception as cleanup_error:
                    logger.error(
                        "memory rebuild cleanup failed after %s: %s",
                        primary_error,
                        cleanup_error,
                    )
                else:
                    try:
                        self._clear_rebuild_marker()
                    except Exception as marker_error:
                        logger.error("memory rebuild marker cleanup failed: %s", marker_error)
                raise primary_error

            if failed_lanes:
                # A non-strict lifecycle rebuild may degrade to any lane that
                # was rebuilt completely.  Exclude failed/partially populated
                # lanes and delete their durable collections so a restart
                # cannot mix incomplete state back in.
                surviving_lanes = [
                    lane for lane in self._lanes if lane.name not in failed_lanes
                ]
                failed_names = {
                    collection_name(self.COLLECTION_NAME, name)
                    for name in failed_lanes
                }
                try:
                    self._delete_collection_names(failed_names)
                except Exception as cleanup_error:
                    logger.error("memory rebuild partial-lane cleanup failed: %s", cleanup_error)
                    self._lanes = []
                    self._collection = None
                    self._healthy = False
                    return False
                self._lanes = surviving_lanes
                self._collection = next(
                    (lane.collection for lane in self._lanes if lane.name == LANE_FASTEMBED),
                    self._lanes[0].collection if self._lanes else None,
                )
                self._healthy = bool(self._lanes)

        try:
            self._clear_rebuild_marker()
        except Exception as e:
            # The index may be complete in this process, but leaving a durable
            # poison marker means it cannot be treated as committed.
            self._healthy = False
            self._lanes = []
            self._collection = None
            logger.error("memory rebuild marker cleanup failed: %s", e)
            if strict:
                raise
            return False

        logger.info(f"MemoryVectorStore rebuilt with {len(ids)} entries across {len(self._lanes)} lanes")
        return self._healthy

    def get_stats(self) -> Dict:
        return {
            "healthy": self.healthy,
            "count": self.count(),
            "lanes": [lane.stats() for lane in self._lanes],
        }

File: rust/src/src/memory_vector.rs
// src/memory_vector.rs  <- src/memory_vector.py
//! SQLite-backed vector store for memory entries.
//!
//! Shares the `EmbeddingClient` with RAG to save memory. Stores pre-computed
//! embeddings (the vector store does not manage embedding).
//!
//! ## Port classification — PORT_VIA_DB (web)
//!
//! The Python `MemoryVectorStore` drives a ChromaDB *collection* object
//! (`"odysseus_memories"`). Per the project infra decision (the deliberate
//! ChromaDB->SQLite swap), this port builds on `crate::src::vector_store`
//! (the `VectorCollection` facade over the shared `vectors` table) and on
//! `crate::src::embeddings::EmbeddingClient`. It is a pure backend swap: the
//! add/remove/count/search/find_similar/rebuild surface is faithful, including
//! the `round(1 - distance, 4)` similarity score and the `0.92` near-duplicate
//! threshold.
//!
//! ### Documented deviations
//!
//! * The embedding client is injected as `Arc<EmbeddingClient>` (mirroring
//!   `app_initializer` sharing one instance across rag/memory/personal-docs) or
//!   lazily built via `get_embedding_client()`, exactly as the Python `__init__`
//!   accepts `embedding_model=None` and falls back to the factory.
//! * `round(x, 4)` uses Python banker's rounding; the Rust port rounds
//!   half-away-from-zero. Cosmetic only — it affects the displayed `score`
//!   digits, not which memory ids rank.
//! * `data_dir` is accepted for signature parity (the Python stores it but never
//!   uses it — ChromaDB persistence is handled by the client, and here by the
//!   shared `vectors` table); it is kept as a field for faithfulness.
//!
//! Feature gate: `web` (HTTP embeddings + the rusqlite-backed vector store).

use std::sync::Arc;

use serde_json::{json, Map, Value};

use crate::error::PyResult;
use crate::pylog as logger;
use crate::src::embeddings::{get_embedding_client, EmbeddingClient};
use crate::src::vector_store::{self, VectorCollection};

/// `COLLECTION_NAME = "odysseus_memories"` (a class attribute in the Python).
pub const COLLECTION_NAME: &str = "odysseus_memories";

/// `round(x, 4)` — half-away-from-zero (documented banker's-rounding drift).
fn round4(x: f64) -> f64 {
    (x * 10_000.0).round() / 10_000.0
}

/// Process-global handle to the live `MemoryVectorStore` (the Python module-level
/// `_memory_vector`). Set at startup from `AppState.memory_vector`; read by the
/// agent-tool path (`ai_interaction::do_manage_memory`) which — unlike the HTTP
/// routes — has no `AppState`. `None` when the vector store is degraded/absent.
static MEMORY_VECTOR: once_cell::sync::Lazy<std::sync::RwLock<Option<Arc<MemoryVectorStore>>>> =
    once_cell::sync::Lazy::new(|| std::sync::RwLock::new(None));

/// Install the global `MemoryVectorStore` (called once at app startup).
pub fn set_memory_vector(mv: Option<Arc<MemoryVectorStore>>) {
    *MEMORY_VECTOR.write().unwrap() = mv;
}

/// Read the global `MemoryVectorStore` (the `if _memory_vector:` check).
pub fn memory_vector() -> Option<Arc<MemoryVectorStore>> {
    MEMORY_VECTOR.read().unwrap().clone()
}

/// `{"source": "memory"}` — the fixed metadata every memory vector carries.
fn memory_meta() -> Value {
    let mut m = Map::new();
    m.insert("source".into(), json!("memory"));
    Value::Object(m)
}

/// One semantic-search result (`{"memory_id": str, "score": float}`).
#[derive(Debug, Clone, PartialEq)]
pub struct MemoryHit {
    pub memory_id: String,
    pub score: f64,
}

impl MemoryHit {
    /// Render as the Python `dict` (a `serde_json::Value` object) for callers
    /// that consume the JSON shape (`hit["memory_id"]`, `hit["score"]`).
    pub fn to_value(&self) -> Value {
        json!({ "memory_id": self.memory_id, "score": self.score })
    }
}

/// Vector index over memory entries for semantic retrieval.
///
/// Mirrors the Python `MemoryVectorStore`. `collection`/`model` are `Option` so
/// the `healthy` flag (`self._healthy`) ports faithfully; the embedding client is
/// shared as `Arc<EmbeddingClient>`.
///
/// `Clone` is cheap and observationally transparent: every field is either a
/// `String`, a shared `Arc<EmbeddingClient>`, or a `VectorCollection` handle
/// (`{ name }`). All real state lives in the shared SQLite `vectors` table, so a
/// cloned store reads/writes the SAME backing collection. This lets a route that
/// holds an `Arc<MemoryVectorStore>` (immutable, shared) hand an owned
/// `&mut Option<MemoryVectorStore>` to `audit_memories` (whose `rebuild` needs
/// `&mut self` to reset the collection handle) — the Python passes the one shared
/// object, and rebuilding the DB-backed collection is identical either way.
#[derive(Clone)]
pub struct MemoryVectorStore {
    #[allow(dead_code)]
    data_dir: String,
    model: Option<Arc<EmbeddingClient>>,
    collection: Option<VectorCollection>,
    healthy: bool,
}

impl MemoryVectorStore {
    /// `MemoryVectorStore.__init__(self, data_dir, embedding_model=None)`.
    ///
    /// Async because `_initialize` may build the embedding client (an HTTP
    /// dimension probe) when none is injected. NEVER raises: an init failure
    /// leaves the instance `unhealthy` (the Python catches and logs).
    pub async fn new(data_dir: &str, embedding_model: Option<Arc<EmbeddingClient>>) -> Self {
        let mut store = MemoryVectorStore {
            data_dir: data_dir.to_string(),
            model: embedding_model,
            collection: None,
            healthy: false,
        };
        store.initialize().await;
        store
    }

    /// `_initialize(self)`.
    ///
    /// If no model was injected, build one via `get_embedding_client()`
    /// (`raise RuntimeError` -> here: log + stay unhealthy when the factory
    /// returns `None`). Open the `"odysseus_memories"` collection and mark
    /// healthy. The whole body is the Python's `try`/`except` — any failure just
    /// leaves `_healthy = False`.
    async fn initialize(&mut self) {
        // if self._model is None: self._model = get_embedding_client()
        if self.model.is_none() {
            match get_embedding_client().await {
                Some(m) => {
                    let m = Arc::new(m);
                    logger::info(&format!("MemoryVectorStore using embeddings: {}", m.url()));
                    self.model = Some(m);
                }
                None => {
                    // raise RuntimeError("No embedding backend available") -> caught.
                    logger::error("MemoryVectorStore init failed: No embedding backend available");
                    return;
                }
            }
        }

        // client.get_or_create_collection(name=..., metadata={hnsw:space:cosine})
        match VectorCollection::get_or_create(COLLECTION_NAME) {
            Ok(c) => {
                let count = c.count().unwrap_or(0);
                self.collection = Some(c);
                self.healthy = true;
                logger::info(&format!("MemoryVectorStore ready (entries={count})"));
            }
            Err(e) => {
                logger::error(&format!("MemoryVectorStore init failed: {e}"));
            }
        }
    }

    /// `healthy` property — `self._healthy`.
    pub fn healthy(&self) -> bool {
        self.healthy
    }

    /// `_embed(self, texts) -> List[List[float]]`.
    ///
    /// `vecs = self._model.encode(texts, normalize_embeddings=True)`. Errors
    /// propagate to the caller (the Python `encode` raises on a down endpoint).
    async fn embed(&self, texts: &[String]) -> PyResult<Vec<Vec<f32>>> {
        let model = self.model.as_ref().expect("embed called without an embedding model");
        model.encode(texts, true).await
    }

    /// `count(self) -> int` — number of stored vectors (`0` when unhealthy).
    pub fn count(&self) -> i64 {
        if !self.healthy {
            return 0;
        }
        self.collection.as_ref().map(|c| c.count().unwrap_or(0)).unwrap_or(0)
    }

    /// `add(self, memory_id, text)` — add a single memory entry to the index.
    ///
    /// Skips when the id already exists (the Python `if existing["ids"]: return`).
    /// A no-op when unhealthy. Errors propagate (the Python lets `encode`/`add`
    /// raise here — only `remove` is wrapped in a `try`).
    pub async fn add(&self, memory_id: &str, text: &str) -> PyResult<()> {
        if !self.healthy {
            return Ok(());
        }
        let collection = self.collection.as_ref().unwrap();
        // existing = collection.get(ids=[memory_id]); if existing["ids"]: return
        let existing = collection.get_by_ids(&[memory_id.to_string()])?;
        if !existing.ids.is_empty() {
            return Ok(());
        }
        let embeddings = self.embed(&[text.to_string()]).await?;
        collection.add(
            &[memory_id.to_string()],
            &embeddings,
            &[text.to_string()],
            &[memory_meta()],
        )?;
        Ok(())
    }

    /// `remove(self, memory_id)` — remove a memory entry (O(1)).
    ///
    /// A no-op when unhealthy. Errors are swallowed with a warning (the Python
    /// `try/except` around the delete).
    pub fn remove(&self, memory_id: &str) {
        if !self.healthy {
            return;
        }
        let collection = self.collection.as_ref().unwrap();
        if let Err(e) = collection.delete(&[memory_id.to_string()]) {
            logger::warning(&format!("memory remove {memory_id}: {e}"));
        }
    }

    /// `search(self, query, k=8) -> List[dict]`.
    ///
    /// Search for the most relevant memory ids by semantic similarity. Returns
    /// `[{"memory_id", "score"}]` where `score = round(1 - distance, 4)`. Empty
    /// when unhealthy or the collection is empty. Errors propagate (the Python
    /// does not wrap this in a `try`).
    pub async fn search(&self, query: &str, k: i64) -> PyResult<Vec<MemoryHit>> {
        if !self.healthy {
            return Ok(Vec::new());
        }
        let collection = self.collection.as_ref().unwrap();
        let count = collection.count()?;
        if count == 0 {
            return Ok(Vec::new());
        }

        let embeddings = self.embed(&[query.to_string()]).await?;
        let query_vec: &[f32] = embeddings.first().map(|v| v.as_slice()).unwrap_or(&[]);
        // actual_k = min(k, count)
        let actual_k = k.min(count);
        let results = collection.query(query_vec, actual_k, None)?;

        let mut out: Vec<MemoryHit> = Vec::new();
        let ids = &results.ids[0];
        // for idx, mid in enumerate(results["ids"][0]):
        for (idx, mid) in ids.iter().enumerate() {
            let distance = results.distances[0][idx];
            out.push(MemoryHit {
                memory_id: mid.clone(),
                score: round4(1.0 - distance),
            });
        }
        Ok(out)
    }

    /// `find_similar(self, text, threshold=0.92) -> Optional[str]`.
    ///
    /// Returns the nearest memory id when its similarity (`1 - distance`) is at
    /// least `threshold`, else `None`. `None` when unhealthy or empty. Errors
    /// propagate (no `try` in the Python).
    pub async fn find_similar(&self, text: &str, threshold: f64) -> PyResult<Option<String>> {
        if !self.healthy {
            return Ok(None);
        }
        let collection = self.collection.as_ref().unwrap();
        if collection.count()? == 0 {
            return Ok(None);
        }

        let embeddings = self.embed(&[text.to_string()]).await?;
        let query_vec: &[f32] = embeddings.first().map(|v| v.as_slice()).unwrap_or(&[]);
        let results = collection.query(query_vec, 1, None)?;

        // if results["ids"][0]: ...
        let ids = &results.ids[0];
        if !ids.is_empty() {
            let distance = results.distances[0][0];
            let similarity = 1.0 - distance;
            if similarity >= threshold {
                return Ok(Some(ids[0].clone()));
            }
        }
        Ok(None)
    }

    /// `rebuild(self, memories)`.
    ///
    /// Rebuild the entire index from a list of memory entries (each a JSON object
    /// with `id` and `text` keys). Drops and recreates the collection, then adds
    /// the entries in batches of 100. A no-op when unhealthy. Errors propagate
    /// (no `try` around the body in the Python).
    pub async fn rebuild(&mut self, memories: &[Value]) -> PyResult<()> {
        if !self.healthy {
            return Ok(());
        }

        // Delete and recreate the collection for a clean rebuild.
        let _ = vector_store::delete_collection(COLLECTION_NAME); // try/except: pass
        let collection = VectorCollection::get_or_create(COLLECTION_NAME)?;

        // Collect (id, text) for entries with both a non-empty stripped text and id.
        let mut texts: Vec<String> = Vec::new();
        let mut ids: Vec<String> = Vec::new();
        for mem in memories {
            // text = mem.get("text", "").strip(); mid = mem.get("id", "")
            let text = mem.get("text").and_then(|v| v.as_str()).unwrap_or("").trim().to_string();
            let mid = mem.get("id").and_then(|v| v.as_str()).unwrap_or("").to_string();
            if !text.is_empty() && !mid.is_empty() {
                texts.push(text);
                ids.push(mid);
            }
        }

        if !texts.is_empty() {
            // Batch in chunks of 100.
            let mut i = 0;
            while i < texts.len() {
                let end = (i + 100).min(texts.len());
                let batch_texts = &texts[i..end];
                let batch_ids = &ids[i..end];
                let embeddings = self.embed(batch_texts).await?;
                let metas: Vec<Value> = (0..batch_ids.len()).map(|_| memory_meta()).collect();
                collection.add(batch_ids, &embeddings, batch_texts, &metas)?;
                i = end;
            }
        }

        self.collection = Some(collection);
        logger::info(&format!("MemoryVectorStore rebuilt with {} entries", ids.len()));
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round4_half_away() {
        assert!((round4(0.123449) - 0.1234).abs() < 1e-12);
        assert!((round4(0.99995) - 1.0).abs() < 1e-12);
    }

    #[test]
    fn memory_meta_shape() {
        assert_eq!(memory_meta(), json!({"source": "memory"}));
    }

    #[test]
    fn memory_hit_to_value() {
        let h = MemoryHit { memory_id: "m1".into(), score: 0.9876 };
        assert_eq!(h.to_value(), json!({"memory_id": "m1", "score": 0.9876}));
    }

    #[test]
    fn unhealthy_store_is_inert() {
        // A store with no collection/model behaves like the Python's
        // `_healthy == False` path: count() == 0 and the mutating ops no-op.
        let store = MemoryVectorStore {
            data_dir: "data".into(),
            model: None,
            collection: None,
            healthy: false,
        };
        assert_eq!(store.count(), 0);
        assert!(!store.healthy());
        store.remove("anything"); // no panic, no-op
    }
}
